reference/behavior_test/establishing_decorator_pattern.rst

```diff
@@ -0,0 +1,369 @@
+Selenium will know which url to go thank to `the protected $path attribute
+<https://github.com/akeneo/pim-community-dev/blob/master/features/Context/Page/Role/Edit.php#L19>`_
+containing in *Context\\Page\\Role\\Edit*.
+
+You may remark that this Gherkin method still use the old implementation of Behat in Akeneo PIM.
+
+Call Pages Actions
+******************
+
+*And I grant rights to group Products* Now, let's jump in Decorators !
+
+All our new contexts are in the namespace *\\Pim\\Behat\\Context\\Domain* which is classified by domain. Each domain is
+linked to a page in Akeneo PIM. That's why you will find in this namespace the same architecture than our main menu.
+This organisation is very helpful to quickly find the context you are looking for.
+
+We are working on permissions localized in the system main menu. So let's see the
+`\\Pim\\Behat\\Context\\Domain\\System\\PermissionsContext <https://github.com/akeneo/pim-community-dev/blob/master/features/Pim/Behat/Context/Domain/System/PermissionsContext.php>`_.
+
+.. image:: images/PermissionsContext.png
+
+.. code-block:: php
+
+    // \Pim\Behat\Context\Domain\System\PermissionsContext
+    /**
+     * @param string $action (grant|remove)
+     * @param string $group
+     *
+     * Gherkin will match this regexp
+     * @When /^I (grant|revoke) rights to groups? (.*)$/
+     */
+    public function iSetRightsToACLGroups($action, $groups) // $action = 'grant', $group = 'Products'
+    {
+        foreach ($this->listToArray($groups) as $group) {
+
+            // In the previous step we put the current page to "Role edit"
+            $page = $this->getCurrentPage(); // Returns an instance of Context\Page\Role\Edit
+            $page->executeActionOnGroup($action, $group);
+        }
+    }
+
+You may notice that our new contexts are really clean. To avoid to pollute context and let them as clean as possible
+the rule is to only call pages method's.
+
+Pages
+_____
+
+The current page is `Context\\Page\\Role\\Edit
+<https://github.com/akeneo/pim-community-dev/blob/master/features/Context/Page/Role/Edit.php>`_.
+
+When you call *Pim\\Behat\\Context\\NavigationContext->getCurrentPage()* the method will match "Role edit" to
+*return new Context\\Page\\Role\\Edit()*. That's why all our pages are classified in a folder with the Entity name and
+then and then each action has its own page like Index.php, Creation.php, Edit.php as you can see in the `Role folder
+<https://github.com/akeneo/pim-community-dev/tree/master/features/Context/Page/Role>`_ for
+example.
+
+Let's follow what happen when calling *executeActionOnGroup()* method :
+
+.. image:: images/edit.png
+
+.. code-block:: php
+
+    /**
+     * Grant or revoke rights to the given specified $group
+     *
+     * @param string $action 'grant'|'revoke'
+     * @param string $group
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function executeActionOnGroup($action, $group) // $action = 'grant', $group = 'Products'
```

Comment:
We should talk about it irl, I dont really understand how we can remove pages. Thanks for the remark.